llvm/lib/CodeGen/LiveIntervals.cpp

```diff
@@ -730,7 +730,16 @@ void LiveIntervals::addKillFlags(const VirtRegMap *VRM) {
     // Find the regunit intervals for the assigned register. They may overlap
     // the virtual register live range, cancelling any kills.
     RU.clear();
-    for (MCRegUnit Unit : TRI->regunits(PhysReg)) {
+    LaneBitmask ArtificialLanes = LaneBitmask::getNone();
+    for (MCRegUnitMaskIterator UI(PhysReg, TRI); UI.isValid(); ++UI) {
+      auto [Unit, Bitmask] = *UI;
+      // Record lane mask for all artificial RegUnits for this physreg.
```

Comment:
Tablegen calculates the "aritificial" property for regunits as well as registers, but I guess we don't have direct access to that here?

https://github.com/llvm/llvm-project/blob/041e5c96c4bd247a3dd6099f49143ee66d9205d8/llvm/utils/TableGen/Common/CodeGenRegisters.h#L550

Reply:
Not directly, unless I generate a table for it. Iterating through the roots shouldn't be much of a problem in practice, since there can never be more than two roots. (It's actually not clear to me why there would ever be more than one root? What's the use-case for this?)

In either case, I'll move this to a special `isArtificial(MCRegUnit)` method to query this info, to avoid having to write this loop in various places in the future. If you think it's worth generating this in a table though, I'm happy to do so!

Reply:
> It's actually not clear to me why there would ever be more than one root? What's the use-case for this?

I think the second root is only ever an aritificial register used to represent ad hoc aliasing (the `Aliases = [...]` field in .td files), but I could be misremembering the details.

Reply:
Looking at ARMGenRegisterInfo.td, I think that's indeed right, thanks for explaining!